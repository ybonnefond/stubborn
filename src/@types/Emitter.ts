export interface Emitter {
  setMaxListeners(nb: number): this;
  getMaxListeners(): number;
  on(event: symbol, listener: (...args: any[]) => void): this;
  once(event: symbol, listener: (...args: any[]) => void): this;
  off(event: symbol, listener: (...args: any[]) => void): this;
}
