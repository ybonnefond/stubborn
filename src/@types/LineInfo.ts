export type LineInfo = {
  file: string;
  line: string;
  column: string;
};
