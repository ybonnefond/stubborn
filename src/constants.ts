export enum METHODS {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export enum STATUS_CODES {
  // 2xx Success
  SUCCESS = 200,

  // 5xx Server Errors
  NOT_IMPLEMENTED = 501,
}

export const EVENTS = {
  NOT_IMPLEMENTED: Symbol.for('NOT_IMPLEMENTED'),
  CLEARED: Symbol.for('CLEARED'),
};

export const WILDCARD = Symbol.for('WILDCARD');

export enum DIFF_TYPES {
  FAIL_EQUALITY = 'FAIL_EQUALITY',
  FAIL_MATCHING = 'FAIL_MATCHING',
  FAIL_FUNCTION = 'FAIL_FUNCTION',
  MISSING = 'MISSING',
  EXTRA = 'EXTRA',
  INVALID_VALUE_TYPE = 'INVALID_VALUE_TYPE',
}
