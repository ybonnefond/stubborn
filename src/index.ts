export * from './@types';
export * from './Stubborn';
export * from './constants';
export * from './Route';
export { logDiffOn501 } from './utils';
