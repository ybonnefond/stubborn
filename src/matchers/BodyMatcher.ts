import { Request } from '../@types';
import { Route } from '../Route';

import { match } from './utils';

export function bodyMatcher(route: Route) {
  return (req: Request) => {
    const { body: rawBody } = req;
    const body = rawBody instanceof Buffer ? String(rawBody) : rawBody;
    const bodyDefinition = route.getBody();

    // Bypass body matching
    if (null === bodyDefinition) {
      return true;
    }

    return match(bodyDefinition, body);
  };
}
