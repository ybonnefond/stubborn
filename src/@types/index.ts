import { IncomingMessage, ServerResponse } from 'http';
import { METHODS, WILDCARD } from '../constants';

export type JsonPrimitive = string | number | boolean | null;
export type JsonObject = { [member: string]: JsonValue };
export interface JsonArray extends Array<JsonValue> {}
export type JsonValue = JsonPrimitive | JsonObject | JsonArray;

export type MatchFunction = (value: JsonValue) => boolean;
export type DefinitionMatcher = RegExp | MatchFunction | typeof WILDCARD;
export type DefinitionValue =
  | DefinitionMatcher
  | JsonValue
  | undefined
  | METHODS; // Any possible values for definition

// Request part definitions types
export type MethodDefinition = METHODS;
export type PathDefinition = DefinitionMatcher | string;
export type HeaderDefinition = DefinitionMatcher | string | number;
export type HeadersDefinition =
  | Record<string, HeaderDefinition>
  | typeof WILDCARD;

export type RequestBodyDefinitionPrimitive = DefinitionMatcher | JsonValue;
export type RequestBodyDefinitionObject = {
  [member: string]: RequestBodyDefinitionValue;
};
export interface RequestBodyDefinitionArray
  extends Array<RequestBodyDefinitionValue> {}
export type RequestBodyDefinitionValue =
  | RequestBodyDefinitionPrimitive
  | RequestBodyDefinitionObject
  | RequestBodyDefinitionArray;
export type RequestBodyDefinition = undefined | RequestBodyDefinitionValue;

export type QueryParameterDefinitionPrimitives =
  | DefinitionMatcher
  | string
  | number;
export type QueryParameterDefinition =
  | QueryParameterDefinitionPrimitives
  | QueryParameterDefinitionPrimitives[];
export type QueryDefinition =
  | Record<string, QueryParameterDefinition>
  | typeof WILDCARD;

export type TemplateFunction = (request: Request, scope: any) => JsonValue;
export type TemplateObject = { [key: string]: Template };
export type TemplateArray = { [index: number]: Template };

export type Template =
  | JsonValue
  | TemplateFunction
  | TemplateObject
  | TemplateArray;

export type ResponseDefinition = {
  statusCode: number;
  headers: Record<string, string | TemplateFunction>;
  body: Template;
};

export type RequestHeaders = Record<string, string>;
export type RequestQuery = Record<string, string | string[]>;
export type RequestBody = JsonValue | undefined;
export type RequestPath = string;
export type RequestHash = string;
export type RequestMethod = string;

export type RequestInfo = {
  headers: RequestHeaders;
  body: RequestBody;
  path: RequestPath;
  query: RequestQuery;
  hash: RequestHash;
  method: RequestMethod;
};

export type Request = IncomingMessage &
  RequestInfo & {
    searchParams?: URLSearchParams;
  };

export type Response = ServerResponse;
export type NextFunction = (err?: any) => void;

export type Middleware = (
  req: Request,
  res: Response,
  next: NextFunction,
) => any;

export type RequestMatcher = (req: Request) => boolean;

export type ResponseHeaders = {
  [key: string]: string;
};

export type ResponseBody = JsonValue;
