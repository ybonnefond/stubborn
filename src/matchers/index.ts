import { bodyMatcher } from './BodyMatcher';
import { headerMatcher } from './HeaderMatcher';
import { methodMatcher } from './MethodMatcher';
import { pathMatcher } from './PathMatcher';
import { queryMatcher } from './QueryMatcher';

export const matchers = {
  headerMatcher,
  methodMatcher,
  pathMatcher,
  queryMatcher,
  bodyMatcher,
};
