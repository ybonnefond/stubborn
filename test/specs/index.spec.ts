import { init } from '../helpers';
import { toReplyWith } from '../matchers';

import { Request } from '../../src/@types';
import { STATUS_CODES } from '../../src/constants';

describe('index', () => {
  expect.extend({ toReplyWith });
  const { sb, request } = init();

  it('should return NOT_IMPLEMENTED if no routes are configured', async () => {
    expect(await request('/')).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
  });

  describe('Request matching', () => {
    describe('Method', () => {
      it('should return NOT_IMPLEMENTED if method does not match', async () => {
        sb.get('/');

        expect(await request('/', { method: 'POST' })).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return SUCCESS if method is GET', async () => {
        sb.get('/');

        expect(await request('/')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if method is POST', async () => {
        sb.post('/');

        expect(await request('/', { method: 'POST' })).toReplyWith(
          STATUS_CODES.SUCCESS,
        );
      });

      it('should return SUCCESS if method is POST with empty body', async () => {
        sb.post('/', '');

        expect(
          await request('/', { method: 'POST', json: false, body: '' }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if method is POST with string body', async () => {
        sb.post('/', 'body');

        expect(
          await request('/', { method: 'POST', json: false, body: 'body' }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if method is PUT', async () => {
        sb.put('/');

        expect(await request('/', { method: 'PUT' })).toReplyWith(
          STATUS_CODES.SUCCESS,
        );
      });

      it('should return SUCCESS if method is PUT with string body', async () => {
        sb.put('/', 'body');

        expect(
          await request('/', { method: 'PUT', json: false, body: 'body' }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if method is PATCH', async () => {
        sb.patch('/');

        expect(await request('/', { method: 'PATCH' })).toReplyWith(
          STATUS_CODES.SUCCESS,
        );
      });

      it('should return SUCCESS if method is PATCH with string body', async () => {
        sb.patch('/', 'body');

        expect(
          await request('/', { method: 'PATCH', json: false, body: 'body' }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if method is DELETE', async () => {
        sb.delete('/');

        expect(await request('/', { method: 'DELETE' })).toReplyWith(
          STATUS_CODES.SUCCESS,
        );
      });
    });

    describe('Path', () => {
      it('should return NOT_IMPLEMENTED if path does not strictly match', async () => {
        sb.get('/test');

        expect(await request('/test2')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return SUCCESS if path strictly match', async () => {
        sb.get('/test');

        expect(await request('/test')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return NOT_IMPLEMENTED if path does not match a regex', async () => {
        sb.get(/\/test\/[0-9]{1}/);

        expect(await request('/test2', { method: 'GET' })).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return SUCCESS if path match regexp', async () => {
        sb.get(/\/test\/[0-9]{1}/);

        expect(await request('/test/2')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if path match custom function', async () => {
        sb.get(path => path === '/test/2');

        expect(await request('/test/2')).toReplyWith(STATUS_CODES.SUCCESS);
      });
    });

    describe('Headers', () => {
      it('should return NOT_IMPLEMENTED if header is not in definition', async () => {
        sb.get('/');

        expect(
          await request('/', { headers: { Authorization: 'token' } }),
        ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return NOT_IMPLEMENTED if header in definition is not provided', async () => {
        sb.get('/').setHeaders({ Authorization: 'token' });

        expect(await request('/')).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return NOT_IMPLEMENTED if headers name are different', async () => {
        sb.get('/').setHeaders({ Authorization: 'token' });

        expect(
          await request('/', { headers: { 'x-api-key': 'api-key' } }),
        ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return NOT_IMPLEMENTED if headers values are different', async () => {
        sb.get('/').setHeaders({ Authorization: 'token' });

        expect(
          await request('/', { headers: { Authorization: 'not-token' } }),
        ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return SUCCESS if header is wildcarded', async () => {
        sb.get('/').setHeaders(null);

        expect(
          await request('/', { headers: { Authorization: 'token' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if header is null and in request', async () => {
        sb.get('/').setHeaders({ Authorization: null });

        expect(
          await request('/', { headers: { Authorization: 'token' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if header is null and not in request', async () => {
        sb.get('/').setHeaders({ Authorization: null });

        expect(await request('/')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if header is strictly equal to the definition', async () => {
        sb.get('/').setHeaders({ Authorization: 'token' });

        expect(
          await request('/', { headers: { Authorization: 'token' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return NOT_IMPLEMENTED if header does not match the definition', async () => {
        sb.get('/').setHeaders({ Authorization: /^Bearer [0-9]{1}$/ });

        expect(
          await request('/', { headers: { Authorization: 'Bearer 11' } }),
        ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return SUCCESS if header match a regexp', async () => {
        sb.get('/').setHeaders({ Authorization: /^Bearer [0-9]{1}$/ });

        expect(
          await request('/', { headers: { Authorization: 'Bearer 9' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if header match a custom function', async () => {
        sb.get('/').setHeaders({
          Authorization: header => header === 'Bearer 9',
        });

        expect(
          await request('/', { headers: { Authorization: 'Bearer 9' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS when setting a specific header', async () => {
        sb.get('/').setHeader('Authorization', 'Bearer 9');

        expect(
          await request('/', { headers: { Authorization: 'Bearer 9' } }),
        ).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should convert json if JSON Accept header', async () => {
        sb.post('/', '{"key": "value"}').setResponseBody({ key: 'ok' });

        const res = await request('/', {
          method: 'post',
          body: '{"key": "value"}',
          json: false,
          headers: { Accept: 'application/vnd.api+json' },
        });
        expect(res).toReplyWith(STATUS_CODES.SUCCESS);

        const body = JSON.parse(res.body);
        expect(body).toStrictEqual({ key: 'ok' });
      });
    });

    describe('Query', () => {
      it('should return NOT_IMPLEMENTED if query parameters is not in definition', async () => {
        sb.get('/');

        expect(await request('/?page=10')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return NOT_IMPLEMENTED if query parameter in definition is not provided', async () => {
        sb.get('/').setQueryParameters({ page: 10 });

        expect(await request('/')).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      it('should return NOT_IMPLEMENTED if query parameters are different', async () => {
        sb.get('/').setQueryParameters({ page: 10 });

        expect(await request('/?limit=5')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return NOT_IMPLEMENTED if query parameter value is not strictly equal to definition', async () => {
        sb.get('/').setQueryParameters({ page: 10 });

        expect(await request('/?page=11')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return SUCCESS if query parameter is strictly equal to the definition', async () => {
        sb.get('/').setQueryParameters({ page: '10' });

        expect(await request('/?page=10')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if query parameters are wildcarded', async () => {
        sb.get('/').setQueryParameters(null);

        expect(await request('/?page=10')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if query parameter is null and in query', async () => {
        sb.get('/').setQueryParameters({ page: null });

        expect(await request('/?page=10')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if query parameter is null and not in query', async () => {
        sb.get('/').setQueryParameters({ page: null });

        expect(await request('/')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if query parameter matches multiple values', async () => {
        sb.get('/').setQueryParameters({ image: ['id1', 'id2'] });

        expect(await request('/?image=id1&image=id2')).toReplyWith(
          STATUS_CODES.SUCCESS,
        );
      });

      it('should return SUCCESS if query parameter does not match multiple values', async () => {
        sb.get('/').setQueryParameters({ image: ['id1', 'id2'] });

        expect(await request('/?image=id1&image=id3')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return NOT_IMPLEMENTED if header does not match the definition', async () => {
        sb.get('/').setQueryParameters({ page: /^\d+$/ });

        expect(await request('/?page=two')).toReplyWith(
          STATUS_CODES.NOT_IMPLEMENTED,
        );
      });

      it('should return SUCCESS if header match a regexp', async () => {
        sb.get('/').setQueryParameters({ page: /^\d+$/ });

        expect(await request('/?page=100')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS if header match a custom function', async () => {
        sb.get('/').setQueryParameters({ page: param => param === '100' });

        expect(await request('/?page=100')).toReplyWith(STATUS_CODES.SUCCESS);
      });

      it('should return SUCCESS when setting a specific query parameter', async () => {
        sb.get('/').setQueryParameter('page', '100');

        expect(await request('/?page=100')).toReplyWith(STATUS_CODES.SUCCESS);
      });
    });

    describe('Body', () => {
      it('should return NOT_IMPLEMENTED if body is not in definition', async () => {
        sb.post('/');

        expect(
          await request('/', { method: 'POST', body: 'test', json: false }),
        ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
      });

      describe('Text body', () => {
        it('should return NOT_IMPLEMENTED if no body in request but in definition', async () => {
          sb.post('/', 'test').setHeaders({ 'Content-Type': 'text/plain' });

          expect(
            await request('/', {
              method: 'POST',
              json: false,
              headers: { 'Content-Type': 'text/plain' },
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });

        it('should return NOT_IMPLEMENTED if body in request but not in definition', async () => {
          sb.post('/').setHeaders({ 'Content-Type': 'text/plain' });

          expect(
            await request('/', {
              method: 'POST',
              body: 'test',
              json: false,
              headers: { 'Content-Type': 'text/plain' },
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });

        it('should return SUCCESS if definition is strictly equal to body', async () => {
          sb.post('/', 'test').setHeaders({ 'Content-Type': 'text/plain' });

          expect(
            await request('/', {
              method: 'POST',
              body: 'test',
              json: false,
              headers: { 'Content-Type': 'text/plain' },
            }),
          ).toReplyWith(STATUS_CODES.SUCCESS);
        });

        it('should return NOT_IMPLEMENTED if string body does not equal definition body', async () => {
          sb.post('/', 'test').setHeaders({ 'Content-Type': 'text/plain' });

          expect(
            await request('/', {
              method: 'POST',
              body: 'test2',
              json: false,
              headers: { 'Content-Type': 'text/plain' },
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });
      });

      describe('Object body', () => {
        it('should return SUCCESS if json body is equal to definition', async () => {
          sb.post('/', { key: 'value' }).setHeaders({
            'Content-Type': 'application/json',
          });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 'value' },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.SUCCESS);
        });

        it('should return NOT_IMPLEMENTED if values are different in object body', async () => {
          sb.post('/', { key: 'value2' }).setHeaders({
            'Content-Type': 'application/json',
          });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 'value' },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });

        it('should return NOT_IMPLEMENTED if keys are different in object body', async () => {
          sb.post('/', { key2: 'value' }).setHeaders({
            'Content-Type': 'application/json',
          });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 'value' },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });

        it('should return NOT_IMPLEMENTED if count keys are different in object definition', async () => {
          sb.post('/', { key: 'value', key2: 'value' }).setHeaders({
            'Content-Type': 'application/json',
          });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 'value' },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
        });

        it('should return SUCCESS if json body is equal to definition with complex object', async () => {
          sb.post('/', {
            subObject: { subArray: [{ key: 'value' }] },
          }).setHeaders({ 'Content-Type': 'application/json' });

          expect(
            await request('/', {
              method: 'POST',
              body: { subObject: { subArray: [{ key: 'value' }] } },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.SUCCESS);
        });

        it('should return SUCCESS if body equals definition with regex', async () => {
          sb.post('/', { key: /\d+/, key2: [{ subkey: /\d+/ }] }).setHeaders({
            'Content-Type': 'application/json',
          });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 11, key2: [{ subkey: 13 }] },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.SUCCESS);
        });

        it('should return SUCCESS if body equals definition with custom function', async () => {
          sb.post('/', {
            key: (val: string) => val === 'test',
            key2: [{ subkey: (val: string) => val === 'toto' }],
          }).setHeaders({ 'Content-Type': 'application/json' });

          expect(
            await request('/', {
              method: 'POST',
              body: { key: 'test', key2: [{ subkey: 'toto' }] },
              json: true,
            }),
          ).toReplyWith(STATUS_CODES.SUCCESS);
        });
      });
    });
  });

  describe('Response', () => {
    it('should respond with NOT_IMPLEMENTED status code and request information', async () => {
      const options = {
        method: 'POST',
        headers: {
          'x-api-key': 'API_KEY',
        },
        body: { key: 'value' },
        json: true,
      };
      expect(await request('/?page=10&limit=100', options)).toReplyWith(
        STATUS_CODES.NOT_IMPLEMENTED,
        {
          message: expect.any(String),
          request: {
            method: 'POST',
            path: '/',
            headers: expect.objectContaining(options.headers),
            query: { page: '10', limit: '100' },
            body: options.body,
          },
        },
      );
    });

    it('should respond with provided status code', async () => {
      const STATUS_CODE = 201;
      sb.get('/').setResponseStatusCode(STATUS_CODE);

      expect(await request('/')).toReplyWith(STATUS_CODE);
    });

    it('should respond with provided headers', async () => {
      sb.get('/').setResponseHeaders({ custom: 'header' });

      expect(await request('/')).toReplyWith(
        STATUS_CODES.SUCCESS,
        '',
        expect.objectContaining({ custom: 'header' }),
      );
    });

    it('should respond with provided header', async () => {
      sb.get('/').setResponseHeader('custom', 'header');

      expect(await request('/')).toReplyWith(
        STATUS_CODES.SUCCESS,
        '',
        expect.objectContaining({ custom: 'header' }),
      );
    });

    it('should respond with a json encoded body if accept header is json', async () => {
      sb.get('/').setResponseBody('toto');

      expect(await request('/')).toReplyWith(STATUS_CODES.SUCCESS, 'toto');
    });

    it('should encode JSON is response content-type header is provided', async () => {
      sb.get('/')
        .setResponseBody('toto')
        .setResponseHeader('Content-Type', 'application/json; charset=utf-8');

      expect(await request('/', { json: false })).toReplyWith(
        STATUS_CODES.SUCCESS,
        JSON.stringify('toto'),
      );
    });

    it('should respond with no encodding if accept and content-type is not provided', async () => {
      sb.get('/').setResponseBody('toto');

      expect(await request('/', { json: false })).toReplyWith(
        STATUS_CODES.SUCCESS,
        'toto',
      );
    });

    it('should respond with provided object body', async () => {
      sb.get('/').setResponseBody({ custom: 'body' });

      expect(await request('/', { json: true })).toReplyWith(
        STATUS_CODES.SUCCESS,
        { custom: 'body' },
      );
    });

    it('should respond with dynamic body', async () => {
      sb.get('/')
        .setQueryParameters({ page: '10' })
        .setResponseBody({
          custom: (req: Request) => req.query.page,
          arr: ['1', '2', () => '3', { one: () => 'un' }],
          sym: Symbol('bol'),
        });

      expect(await request('/?page=10')).toReplyWith(STATUS_CODES.SUCCESS, {
        custom: '10',
        arr: ['1', '2', '3', { one: 'un' }],
        sym: 'bol',
      });
    });

    it('should pass the current scope for deep object template', async () => {
      sb.get('/')
        .setQueryParameters({ page: '22' })
        .setResponseBody({
          custom: {
            val1: 'some value',
            val2: (req: Request, scope: any) =>
              `${scope.val1} - page: ${req.query.page}`,
          },
        });

      expect(await request('/?page=22')).toReplyWith(STATUS_CODES.SUCCESS, {
        custom: {
          val1: 'some value',
          val2: 'some value - page: 22',
        },
      });
    });
  });

  describe('README examples', () => {
    it('should respond to query', async () => {
      const body = { some: 'body' };
      sb.get('/').setResponseBody({ some: 'body' });

      expect(
        await request('/', {
          json: true,
        }),
      ).toReplyWith(STATUS_CODES.SUCCESS, body);
    });

    it('should respond 501 if a parameter is missing', async () => {
      sb.get('/').setQueryParameters({ page: '1' });

      expect(
        await request('/', {
          throwHttpErrors: false,
        }),
      ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
    });

    it('should respond 501 if a parameter is added', async () => {
      sb.get('/').setQueryParameters({ page: '1' });

      expect(
        await request('/?page=1&limit=10', {
          throwHttpErrors: false,
        }),
      ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
    });

    it('should respond 501 if a parameter is not equal', async () => {
      sb.get('/').setQueryParameters({ page: '1' });

      expect(
        await request('/?page=2', {
          throwHttpErrors: false,
        }),
      ).toReplyWith(STATUS_CODES.NOT_IMPLEMENTED);
    });

    it('should respond using wildcard for parameter and headers', async () => {
      sb.get('/')
        .setQueryParameters({ page: null })
        .setHeaders(null);

      expect(
        await request('/?page=2', {
          headers: { 'x-api-key': 'api key', 'any-other-header': 'stuff' },
          throwHttpErrors: false,
        }),
      ).toReplyWith(STATUS_CODES.SUCCESS);
    });

    it('should match using regex', async () => {
      sb.post('/', {
        slug: /^[a-z\-]*$/,
      })
        .setQueryParameters({ page: /^\d$/ })
        .setHeaders(null);

      const res = await request('/?page=2', {
        method: 'POST',
        body: { slug: 'stubborn-ws' },
      });

      expect(res).toReplyWith(STATUS_CODES.SUCCESS);
    });

    it('should match using a function', async () => {
      sb.get('/').setQueryParameters({
        page: value => parseInt(value as string) > 0,
      });

      const res = await request('/?page=2');

      expect(res).toReplyWith(STATUS_CODES.SUCCESS);
    });
  });

  describe('Retain Calls', () => {
    it('should have no calls when initialize the route', () => {
      const route = sb.get('/');
      expect(route.countCalls()).toBe(0);
    });

    it('should retain every calls made to the route', async () => {
      const route = sb.get('/');

      const TOTAL_CALLS = 3;

      for (let i = 1; i <= TOTAL_CALLS; i++) {
        await request();
        expect(route.countCalls()).toBe(i);
      }
    });

    it('should retain a striped version of the request', async () => {
      const route = sb
        .post('/', null)
        .setHeaders(null)
        .setQueryParameters(null);

      await request('/?page=100', { method: 'POST', body: { some: 'body' } });

      expect(route.countCalls()).toBe(1);
      expect(route.getCall(0)).toEqual({
        method: 'POST',
        query: { page: '100' },
        body: { some: 'body' },
        path: '/',
        hash: '',
        headers: {
          accept: 'application/json',
          'accept-encoding': expect.any(String),
          connection: expect.any(String),
          'content-length': expect.any(String),
          'content-type': expect.any(String),
          host: expect.any(String),
          'user-agent': expect.any(String),
        },
      });
    });
  });
});
