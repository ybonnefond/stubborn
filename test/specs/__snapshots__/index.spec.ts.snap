// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index debug should output various stuff 1`] = `
"Method
- Received: post
+ Expected: put

Path
- Received: /test
+ Expected: /

Headers
+   x-header-missing: missing-header-123
-   x-header-extra: x-extra-header-123

-   x-header-1: Bearer world
+   x-header-1: Bearer hello

Query
+   missingParam: /^[0-9]+$/
-   extraParam: 10

-   param1: ten
+   param1: /^[0-9]+$/

Body
+   missingKey: key missing
-   extraKey: extra key value

-   pets.0.type: dog
+   pets.0.type: cat

-   firstname: 123
+   firstname: /^[a-z]+$/

-   lastname: Donald
+   lastname: val => val === 'Doe'

-   roles: writer
+   roles: [writer,reviewer]"
`;
